Structure log update

jacob-toller opened the pull request:
- Structure log has been refactored such that it will 'gather' all the node information first, and then format it at the end
- Tests present as a 'full integration' test from both code and commandline entrypoints.
- Could do with more unit tests but lack decent test data
- Improved actual logging capability of the tool to clearly display conduit length and use of replicate units.

scleakey commented on floodmodeller_api/test/test_toolbox_structure_log.py:
```diff
@@ -79,156 +83,110 @@ def structure():
-    assert output == ["Mannings: 0", "h: 0.00 x w: 0.00"]
+@pytest.fixture
+def ex18_dat_expected():
+    # I think this is the correct way to establish the same 'expected' result, but if this is a larger/more structure-rich dat, could read from file?
```
I think this is okay for such a small dat, but if it was much bigger, then I would read it from a file

scleakey commented on floodmodeller_api/test/test_toolbox_structure_log.py:
```diff
@@ -79,156 +83,110 @@ def structure():
 
 
-def test_add_conduits(slb, conduit_filled, tmpdir):
+def test_conduit_data(slb, conduit_empty):
```
I would maybe rename this to something like `test_empty_conduit` so it's more descriptive

jacob-toller replied:
addressed

scleakey commented on floodmodeller_api/test/test_toolbox_structure_log.py:
```diff
@@ -79,156 +83,110 @@ def structure():
+
+    slb._dat = conduit_chain_dat
+    tmp_csv = Path(tmpdir) / "test_multi_conduits.csv"
+    print(tmp_csv)
```
I would remove the `print` statement before it gets merged

jacob-toller replied:
addressed

scleakey commented on floodmodeller_api/test/test_toolbox_structure_log.py:
```diff
@@ -79,156 +83,110 @@ def structure():
+def test_full_dat_from_commandline(slb, tmpdir, ex18_dat_path, ex18_dat_expected):
+    # these two tests should be as described in the toolbox documentation
+    tmp_csv = Path(tmpdir) / "test_full_dat_from_python.csv"
+    subprocess.call(f'fmapi-structure_log --input_path "{ex18_dat_path}" --output_path "{tmp_csv}"')
```
Nice that you added in this test!

scleakey commented on floodmodeller_api/test/test_toolbox_structure_log.py:
```diff
@@ -79,156 +83,110 @@ def structure():
+    assert text == ex18_dat_expected
+
+
+if __name__ == "__main__":
```
I would get rid of this before merging

jacob-toller replied:
addressed

scleakey commented on floodmodeller_api/test/test_toolbox_structure_log.py:
```diff
@@ -79,156 +83,110 @@ def structure():
+def test_full_dat_from_python(slb, tmpdir, ex18_dat_path, ex18_dat_expected):
+    # these two tests should be as described in the toolbox documentation
+
+    # Im not sure if this is slightly redundant compared to test from commandline
```
I understand what you mean! I think it's worth keeping it though as it might help with diagnosing problems if ever something got broken e.g. if `test_full_dat_from_commandline` was failing but this one wasn't, then we'd immediately know the problem was with the command line part.

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -272,16 +514,72 @@ def _write(  # noqa: PLR0913
-            self._add_conduits()
 
-            self._add_structures()
+if __name__ == "__main__":
```
I would remove this before merging, but if it is an informal test, then you can turn it into a test if you want

jacob-toller replied:
deleted and moved to off-repo file

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -1,14 +1,48 @@
         self.csv_output_path = output_path
+        self.conduit_chains = (
+            {}
+        )  # pylint flags these for type hinting, but not sure how to specify, to discuss
```
Is this just an empty dictionary? (i.e. not sure if you were trying to define a tuple of dictionaries, as that would require an extra comma). If I remember correctly, mypy likes you to typehint a dictionary when defining it so we know what type of dictionary it is. So you would write something like e.g. `self.conduit_chains: dict[str, list[str]] = {}` if it was destined to be a dictionary of lists of strings

jacob-toller replied:
I did add this exact typehinting in a later commit, the parenthesis and new line was something that black insisted on adding.

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -1,14 +1,48 @@
+        self.already_in_chain = set()
 
-    def _add_fields(self):
+        self.unit_store = {}
```
same with this one needing a typehint I imagine

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -19,237 +53,445 @@ def _add_fields(self):
-    def _circular_data(self, conduit, length):
-        dimensions = f"dia: {conduit.diameter:.2f} x l: {length:.2f}"
-        all_mannings = [
+        # check if the
```
?

jacob-toller replied:
addressed, and added/clarified other comments in the area

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -19,237 +53,445 @@ def _add_fields(self):
+            "friction_eq": "COLEBROOK-WHITE",
+            "friction_set": friction_set,
+            "all_friction": all_friction,
+        }  # friction should be CW here; should test this
```
did you test it?

jacob-toller replied:
Yes, FM only allows for colebrook-white friction for section conduits; although I did possibly identify an issue with the default conduit parameters in the API, related to this

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -19,237 +53,445 @@ def _add_fields(self):
+        )
+
+        opening_data = []
+        if (
```
I think the mix of comments and the black formatter is making this look a bit weird - I would move the comments onto their own line so black doesn't try to split all these if statements & loops into multiple lines, if that makes sense?

jacob-toller replied:
addressed, hopefully black doesnt butcher it again

scleakey replied:
I don't think this is addressed, unless you haven't pushed this change?

jacob-toller replied:
should be fixed now

scleakey reviewed (changes requested):
Nice job capturing the behaviour in tests before refactoring! Also I liked how the methods were quite small and your comments were helpful explanations about why the code's doing certain things.

Aside from the line-specific comments, there are some general comments:
- In the tests, it would be better to use `tmp_path` rather than turning `tmpdir` into a `Path`
- In the code and optionally the tests, could you please put in typehints to help out future developers
- In the code, could you make sure any docstrings are more formally written i.e. with capital letters and apostrophes where required
- Could you add a little `TODO` marker in front of anywhere where you say it requires more work, so that those ideas don't get lost
- Try running pytest-cov to see which areas of the code have low coverage (it currently has 63% coverage and I think we'd want it higher than that)

I am going to have a closer look at how the code actually works later, this is just all I could do while on a train!

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -1,14 +1,46 @@
 from floodmodeller_api import DAT
 
 
+def extract_attrs(object_input, keys):
```
I like that you extracted this as a function - I think it helps with readability

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -1,14 +1,46 @@
+    Replaces tuple keys with string versions of themselves, wrapped in ()
+
+    """
+    new_dict = {}
```
Can you double check that you can't write `return {f"({','.join(key)})": value for key, value in old_dict.items()}` here instead (making sure to use that combination of single and double quotes)? I think this would fail a new linting check we just added to the main branch yesterday

jacob-toller replied:
ammended, I think this is slightly different than what I originally had as this one doesnt break the linting.

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -1,14 +1,46 @@
 class StructureLogBuilder:
-    def __init__(self, input_path, output_path) -> None:
+
+    def __init__(self, input_path="", output_path="") -> None:
```
Why are the defaults for each argument `""` - is it helpful?

jacob-toller replied:
This was to allow the testing scenarios and the checker stuff to initialise an instance of the SLB with blank input/output paths and 'manually feed' it, in all fairness None could have been better

scleakey replied:
I think this is one where we should fix it as it's not just the internal workings of a class but its interface - could you please change the defaults to `None`?

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -19,237 +51,446 @@ def _add_fields(self):
+        return conduit_data, add_to_conduit_stack
+
+    def _circular_data(self, conduit):
+        dimensions = extract_attrs(conduit, {"diameter", "invert"})
```
It wouldn't make much of a difference here due to the size (more of a matter of general good practice) but it's more efficient to iterate through lists `["diameter", "invert"]` than through sets `{"diameter", "invert"}` - sets are better for checking membership. This happens in a few places.

jacob-toller replied:
I used sets here as the entries have to be unique, although Ive not heard the difference between iterating through lists and sets before

scleakey commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -19,237 +51,446 @@ def _add_fields(self):
-    def _sprungarch_data(self, conduit, length):
-        dimensions = f"(Springing: {conduit.height_crown:.2f}, Crown: {conduit.height_springing:.2f}) x w: {conduit.width:.2f} x l: {length:.2f}"
-        all_mannings = [
+    def _sprungarch_data(self, conduit):
```
This is a replica of `_sprung_data` - to discuss

scleakey replied:
@joe-pierce any thoughts?

scleakey reviewed (changes requested):
I had a further look on another train!

Some further general comments:
- I am not sure the fixture `slb` is very helpful. Surely it would be better to initialise it with the correct paths or just not have it as a fixture at all?
- Perhaps as the last thing you do, could you just check all the comments that still exist in the code, and if there are any that are notes to me or you but not useful to anyone else, delete them?

jacob-toller commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -19,237 +51,446 @@ def _add_fields(self):
+                    row["Finish"] - row["Start"]
+                )  # round here because fp precision looks uncool
+
+                # this is crude and can be wrong where the min is on the 'edge' of the section instead of at one of the points, but should be correct 99.9% of the time.
```
Comment removed because of this worked on an assumption that the bridge opening points could be anywhere, while FM only allows opening start/end points on section datapoints.

jacob-toller commented on floodmodeller_api/toolbox/model_build/structure_log/structure_log.py:
```diff
@@ -19,237 +51,446 @@ def _add_fields(self):
+
+        return {"dimensions": dimensions}
 
     def _add_structures(self):
```
Cant remember if I commented this elsewhere, this could probably do with a refactor and would also help with https://github.com/People-Places-Solutions/floodmodeller-api/pull/141/files#r1831033444 , although it might be better if we leave as-is for now, and register it as an issue to be handled later

jacob-toller commented:
> In the tests, it would be better to use tmp_path rather than turning tmpdir into a Path
addressed

> In the code and optionally the tests, could you please put in typehints to help out future developers
addressed, no idea why mypy wasnt originally flagging this as an issue.

> In the code, could you make sure any docstrings are more formally written i.e. with capital letters and apostrophes where required
addressed

> Could you add a little TODO marker in front of anywhere where you say it requires more work, so that those ideas don't get lost
done, will also add them as issuesonce we're ready to merge to main

> Try running pytest-cov to see which areas of the code have low coverage (it currently has 63% coverage and I think we'd want it higher than that)
currently at 65% for structure_log.py and 86% total

> I am not sure the fixture slb is very helpful. Surely it would be better to initialise it with the correct paths or just not have it as a fixture at all?
I had this as a fixture from the legacy code, and kept it incase any changes have to be made to the initialisation to reduce redundant code.

> Perhaps as the last thing you do, could you just check all the comments that still exist in the code, and if there are any that are notes to me or you but not useful to anyone else, delete them?
Should be done, but no doubt I've missed one or two.

jacob-toller commented:
All automated checks have been passed, testing regime now includes multiple dat files which should have a greater coverage

joe-pierce commented:
I've run this on some other DAT files and all working as expected. Only small issue is where comments in a unit get poorly interpreted by excel as a formula/weird data type, but this is more of an excel issue and can be resolved by importing the csv as a table rather than directly opening. 

Otherwise happy to merge
